### DISCUSSION TITLE
Suggestion: Use TypeScript?

### DISCUSSION THREAD

**Discussion by r002** [mithi/epic-react-exercises#127] [Ideas]
Hi @mithi !  Thank you again so much with sharing your code online!  I've truly learned a ton from just reading how you wrote everything.

I've been learning React myself and as I've gotten more comfortable with it these past few weeks, I revisited your codebase.  One thing I noticed, compared to other examples I've found online, is that you seem to prefer JavaScript over TypeScript?  I noticed that for both this project and your other `Todo App` (good job!  It's looks great! 👍), you chose vanilla JS.  I guess I was just wondering, is there a reason?

I remember two weeks ago when I first looked at your code, I was initially confused because you often use the spread operator to expand props like `...style`.  So, something like this:

https://github.com/mithi/epic-react-exercises/blob/3a44e66ceb7f073cc0d391390e5dc0d53222d5df/components/button/index.js#L119-L128

But as a total outsider seeing your code for the first time, it wasn't immediately apparent to me what was inside `...style`.  I remember I needed to actually use Chrome's Inspector to see what the props actually were.

For my own projects, I've been exclusively using TypeScript.  Yes, it's definitely more writing and initial work up front but there's that old saying about **_"Code is read far more often than it's written."_** (someone famous said that!  Fowler, I think) so I feel the upfront investment is worth it.  For example, in my own toy React app, I recently coded a "ThemeToggler" button that actually reminded me of your code. 😄 

My simple ThemeToggler button looks like this:
```ts
      <ThemeToggler label='☀️'
                    margin='0 15px 0 0' />
```

and then in my impl, I could either do this:
```ts
type TThemeToggler = {
  margin: string,
  label: string
}

const ThemeToggler = ({ margin, label }: TThemeToggler) => {
...
```
or just:
```ts
type TThemeToggler = {
  margin: string,
  label: string
}

const ThemeToggler = (props: TThemeToggler) => {
...
```

And even if your type definitions are squirreled away and imported from elsewhere, in VSCode, hovering over types instantly very **_explicitly_** tells a coder what a specific function (pure function, React widget, etc) is expecting:

![image](https://user-images.githubusercontent.com/45280066/113005128-318b1680-9142-11eb-870d-6f372210488b.png)

I think like you said before about the "CSS in JS" (you've totally convinced me, btw!  I've become a huge fan of Styled Components! 😀) , this is entirely a matter of preference.  But that's one suggestion I would give to you.  **_Explicit is better than implicit._**  For your next project or if you're just looking to learn/use something different, maybe consider TypeScript?

Anyway, keep up the great work!  Love your code!  ❤️

**Comment by mithi** [comment]
@r002 

Hey! Nice to hear from you again. 

No, I don't prefer javascript over typescript. I'm just faster in javascript than typescript 😅 

The ToDo app was the coding exam given to me by one of the companies I was applying for, and I was given only three days, 
so I decided to write it in Javascript 😅 

For this project, I wrote it in javascript, because at the start I felt like spending too much time configuring typescript stuff with NextJS and fighting the static type checker, and it was not fun.

Of course this is largely due to my incompetence 😅  I'm not that good at typescript but I have since gotten better (thanks to https://react-typescript-cheatsheet.netlify.app/docs/basic/troubleshooting/types). 
Btw I also recently wrote couple of notes about Typescript [here](https://github.com/mithi/epic-react-exercises/wiki), you might find it useful.

I disagree with your argument with `style` though because `style` is a built-in prop to any html component, and it's of type `CSSProperties` (when you use typescript), that's why I think it's safe to say that it's universally known what it is (just like `className`) 

Btw, I want to clarify, inline styling (ie using the `style` prop, is generally frowned upon because of performance reasons (although in my (controversial??) opinion, if you can't prove there is a performance difference, then it's probably not worth optimizing)... but for "real", non-side project applications, it would probably be best if if you use css-in-js libraries like what you mentioned (`styled component`)


Why do you like styled component more than emotion by the way? ❤️

**Reply by r002** [reply]
Hi @mithi !  Thank you for replying so quickly!  Sorry it's taken me so long to reply.  That's incredible that you'd built the ToDo app in only three days though! 😮  I hope the job application went well!  🤞

Yeah, I've had the same initial challenges fighting TS's static type checking too.  Though, for me, I've been finding that it forces me to slow down and think about my app design more carefully, which I think is good for me.  Or at least what I need right now.  Thank you for sharing those notes though!  Bookmarked! ❤️ 

I've actually never used emotion so I don't know!  Haha, I picked up styled-components because it happened to be at the top of some random "best-CSS-in-JS" article that I'd been reading and appeared to be used by many big name mega-corps.  Am still learning them but basically I love just having everything in a single file and also being able to define my base elements as styled components and then reuse those elements (like a `button`, for example) in my React functional components (`Login button`, `Logout button`, etc).  The syntax of SCs also then allow me to set CSS properties via parameters as well.

So, for example, right now I do stuff like:

![image](https://user-images.githubusercontent.com/45280066/113651872-ccaf4f00-9660-11eb-929f-f3bf71523db6.png)

and then:

![image](https://user-images.githubusercontent.com/45280066/113651937-e6509680-9660-11eb-8a85-1223b4aa3ad6.png)

On a completely different note though, btw, I actually wrote about your project and my own experience contributing to it last month!  It was just such a positive and wonderful experience that I wanted to share with everyone!  😀  I just posted my article earlier tonight to the CodeNewbie Forem site; here's the link: [Project #1 | March 6, 2021: mithi/epic-react-exercises](https://community.codenewbie.org/r002/project-1-march-6-2021-mithi-epic-react-exercises-3n2g).

Thank you again so much for sharing your project and for being so generous with your time! 🙏 I basically just told everyone in the article how great your project is and encouraged them to all contribute to it!  😄🚀🎉

**Reply by mithi** [reply]
> On a completely different note though, btw, I actually wrote about your project and my own experience contributing to it last month! It was just such a positive and wonderful experience that I wanted to share with everyone! 😀 I just posted my article earlier tonight to the CodeNewbie Forem site; here's the link: Project #1 | March 6, 2021: mithi/epic-react-exercises.

Hey, this is so nice. I was feeling a bit down yesterday, so this cheered me up a little bit. Thanks for sharing. 

My [bare minimum 3d](https://github.com/mithi/bare-minimum-3d) package is written in Typescript. You might find it interesting.

I copy pasted Josh W Comeau's thoughts about CSS-in-JS libraries and their performance [here](https://github.com/mithi/epic-react-exercises/wiki/CSS-in-JS-solutions-and-Performance). You might find it useful. Enjoy!
